cpp/arcticdb/column_store/key_segment.cpp

```diff
@@ -0,0 +1,146 @@
+                } else {
+                    symbol_ = safe_convert_to_numeric_id(stream_ids_->template reference_at<uint64_t>(0));
+                }
+            }
+            break;
+        case SymbolStructure::UNIQUE:
+            debug::check<ErrorCode::E_ASSERTION_FAILURE>(check_symbols_all_unique(),
+                                                         "Expected all symbols to be unique in KeySegment");
+        case SymbolStructure::UNKNOWN:
+        default:
+            break;
+    }
+}
+
+std::variant<std::vector<AtomKeyPacked>, std::vector<AtomKey>> KeySegment::materialise() const {
+    auto version_data = version_ids_->data();
+    auto creation_ts_data = creation_timestamps_->data();
+    auto content_hash_data = content_hashes_->data();
+    auto index_start_data = start_indexes_->data();
+    auto index_end_data = end_indexes_->data();
+    auto key_types_data = key_types_->data();
+
+    auto version_it = version_data.template cbegin<uint64_TDT>();
+    auto creation_ts_it = creation_ts_data.template cbegin<int64_TDT>();
+    auto content_hash_it = content_hash_data.template cbegin<uint64_TDT>();
+    auto key_types_it = key_types_data.template cbegin<uint8_TDT>();
+
+    if (symbol_structure_ == SymbolStructure::SAME && !is_sequence_type(start_indexes_->type().data_type())) {
+        // Can return AtomKeyPacked vector, much more time and memory efficient, and most common case
+        std::vector<AtomKeyPacked> res;
+        res.reserve(num_keys_);
+        auto index_start_it = index_start_data.template cbegin<int64_TDT>();
+        auto index_end_it = index_end_data.template cbegin<int64_TDT>();
+        for (size_t row_idx = 0;
+             row_idx < num_keys_;
+             ++row_idx, ++version_it, ++creation_ts_it, ++content_hash_it, ++index_start_it, ++index_end_it, ++key_types_it) {
+            res.emplace_back(*version_it, *creation_ts_it, *content_hash_it, KeyType(*key_types_it), *index_start_it, *index_end_it);
+        }
+        return res;
+    } else {
+        // Fall back to returning fully materialised AtomKeys
+        std::vector<AtomKey> res;
+        res.reserve(num_keys_);
+        auto stream_id_data = stream_ids_->data();
+        auto stream_id_it = stream_id_data.template cbegin<uint64_TDT>();
+        const bool stream_id_is_sequence_type = is_sequence_type(stream_ids_->type().data_type());
+        const bool index_is_sequence_type = is_sequence_type(start_indexes_->type().data_type());
+        auto index_start_it = index_start_data.template cbegin<int64_TDT>();
+        auto index_end_it = index_start_data.template cbegin<int64_TDT>();
+        for (size_t row_idx = 0;
+             row_idx < num_keys_;
+             ++row_idx, ++stream_id_it, ++version_it, ++creation_ts_it, ++content_hash_it, ++index_start_it, ++index_end_it, ++key_types_it) {
+            // String columns are stored as uint64s (offsets into the string pool), but StringPool::get_const_view expects an int64
```

Comment:
Can we change that? That seems quite weird and wrong

Reply:
Decided not to change as strong none/nan are defined in terms of `std::numeric_limits<int64_t>::max()`